core/src/main/java/org/infinispan/persistence/support/ComposedSegmentedLoadWriteStore.java

```diff
@@ -120,30 +121,38 @@ public int size() {
    }
 
    @Override
-   public Flowable<K> publishKeys(IntSet segments, Predicate<? super K> filter) {
-      return PersistenceUtil.parallelizePublisher(segments, scheduler, i -> {
+   public Publisher<K> publishKeys(IntSet segments, Predicate<? super K> filter) {
+      IntFunction<Publisher<K>> publisherFunction = i -> {
          AdvancedLoadWriteStore<K, V> alws = stores.get(i);
          if (alws != null) {
             return alws.publishKeys(filter);
          }
          return Flowable.empty();
-      });
+      };
+      if (segments.size() == 1) {
+         return publisherFunction.apply(segments.iterator().nextInt());
+      }
+      return PersistenceUtil.parallelizePublisher(segments, scheduler, publisherFunction);
    }
 
    @Override
-   public Flowable<K> publishKeys(Predicate<? super K> filter) {
+   public Publisher<K> publishKeys(Predicate<? super K> filter) {
       return publishKeys(IntSets.immutableRangeSet(stores.length()), filter);
    }
 
    @Override
    public Publisher<MarshallableEntry<K, V>> entryPublisher(IntSet segments, Predicate<? super K> filter, boolean fetchValue, boolean fetchMetadata) {
-      return PersistenceUtil.parallelizePublisher(segments, scheduler, i -> {
+      IntFunction<Publisher<MarshallableEntry<K, V>>> publisherFunction = i -> {
          AdvancedLoadWriteStore<K, V> alws = stores.get(i);
          if (alws != null) {
             return alws.entryPublisher(filter, fetchValue, fetchMetadata);
          }
          return Flowable.empty();
-      });
+      };
+      if (segments.size() == 1) {
+         return publisherFunction.apply(segments.iterator().nextInt());
+      }
+      return PersistenceUtil.parallelizePublisher(segments, scheduler, publisherFunction);
```

Comment:
Other operations? These are running on the blocking thread pool. Non blocking operations will work just fine. Also since we don't specify a parallelism level in the `parallelizePublisher` method it will use at most `numCPU` blocking threads per call. So if CPU is 4 core and blocking thread pool is 50, you would still have 42 blocking threads that are idle.

Reply:
Key-based operations work with the same stores, so if a bulk operation needs to block a thread a key operation will also need to block a thread. And while some users will indeed run with 4 cores and 50 blocking threads, I believe users with 12 cores x2 with hyperthreading enabled will also want to run with 50 blocking threads.

Reply:
Well if they have 12 cores and HT they will definitely not want to be running with only 50 blocking threads. This is one of the things I have been toying with in the back of my mind is that it seems like # of blocking threads default should also scale based on num cores most likely. Note sure if that is 5 per or what with a floor of 10, haven't really thought it out much.

Reply:
I'm not convinced the # of blocking threads should scale with the # of CPUs, at the very least it should not scale linearly IMO. E.g. with a JDBC store, it doesn't matter how many CPUs you have, I don't think you'd want 200 simultaneous connections to the DB from each node.

Reply:
I am not proposing that in this PR btw, so this should not be holding up any integration of said work. We should discuss this in https://issues.jboss.org/browse/ISPN-10309 or one of its subtasks.

Also just an FYI, but it is currently possible to have more connections than that to the DB with 9.4.x on just the default thread pool sizes.